Sherlock/replicator.h

```diff
@@ -0,0 +1,296 @@
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef CURRENT_SHERLOCK_REPLICATOR_H
+#define CURRENT_SHERLOCK_REPLICATOR_H
+
+#include <functional>
+#include <string>
+#include <thread>
+
+#include "exceptions.h"
+#include "sherlock.h"
+#include "stream_data.h"
+
+#include "../Blocks/HTTP/api.h"
+#include "../Blocks/SS/ss.h"
+
+#include "../Bricks/sync/scope_owned.h"
+#include "../Bricks/waitable_atomic/waitable_atomic.h"
+
+#include "../TypeSystem/Reflection/types.h"
+
+namespace current {
+namespace sherlock {
+
+template <typename STREAM_ENTRY>
+class SubscribableRemoteStream final {
+ public:
+  using entry_t = STREAM_ENTRY;
+
+  class RemoteStream final {
+   public:
+    RemoteStream(const std::string& url, const std::string& top_level_name, const std::string& namespace_name)
+        : url_(url),
+          schema_(Value<reflection::ReflectedTypeBase>(reflection::Reflector().ReflectType<entry_t>()).type_id,
+                  top_level_name,
+                  namespace_name) {}
+
+    void CheckSchema() const {
+      const auto response = HTTP(GET(url_ + "/schema.simple"));
+      if (static_cast<int>(response.code) == 200) {
+        const auto remote_schema = ParseJSON<SubscribableSherlockSchema>(response.body);
+        if (remote_schema != schema_) {
+          CURRENT_THROW(RemoteStreamInvalidSchemaException());
+        }
+      } else {
+        CURRENT_THROW(RemoteStreamDoesNotRespondException());
+      }
+    }
+
+    const std::string GetURLToSubscribe(uint64_t index) const {
+      return url_ + "?i=" + current::ToString(index);
+    }
+
+    const std::string GetURLToTerminate(const std::string& subscription_id) {
+      return url_ + "?terminate=" + subscription_id;
+    }
+
+   private:
+    const std::string url_;
+    const SubscribableSherlockSchema schema_;
+  };
+
+  template <typename F, typename TYPE_SUBSCRIBED_TO>
+  class RemoteSubscriberThread final : public current::sherlock::SubscriberScope::SubscriberThread {
+   public:
+    RemoteSubscriberThread(ScopeOwned<RemoteStream>& remote_stream,
+                           F& subscriber,
+                           uint64_t start_idx,
+                           std::function<void()> done_callback)
+        : valid_(false),
+          done_callback_(done_callback),
+          external_stop_(false),
+          remote_stream_(remote_stream, [this]() { TerminateSubscription(); }),
+          subscriber_(subscriber),
+          index_(start_idx),
+          thread_([this]() { Thread(); }) {
+      valid_ = true;
+    }
+
+    ~RemoteSubscriberThread() {
+      if (valid_) {
+        if (!subscriber_thread_done_) {
+          TerminateSubscription();
+        }
+        thread_.join();
+      } else {
+        if (done_callback_) {
+          done_callback_();
+        }
+      }
+    }
+
+   private:
+    static_assert(current::ss::IsEntrySubscriber<F, TYPE_SUBSCRIBED_TO>::value, "");
```

Comment:
Nit: a `static_assert` on a class-level invariant better be the first line of the body of the lass.